database/rpcdb/db_client.go

```diff
@@ -189,16 +186,80 @@ type iterator struct {
 	db *DatabaseClient
 	id uint64
 
-	data []*rpcdbpb.PutRequest
-	errs wrappers.Errs
+	data        []*rpcdbpb.PutRequest
+	fetchedData chan []*rpcdbpb.PutRequest
+
+	errLock sync.RWMutex
+	err     error
+
+	reqUpdateError chan chan struct{}
+
+	once     sync.Once
+	onClose  chan struct{}
+	onClosed chan struct{}
+}
+
+func newIterator(db *DatabaseClient, id uint64) *iterator {
+	it := &iterator{
+		db:             db,
+		id:             id,
+		fetchedData:    make(chan []*rpcdbpb.PutRequest),
+		reqUpdateError: make(chan chan struct{}),
+		onClose:        make(chan struct{}),
+		onClosed:       make(chan struct{}),
+	}
+	go it.fetch()
+	return it
+}
+
+func (it *iterator) fetch() {
+	defer func() {
+		resp, err := it.db.client.IteratorRelease(context.Background(), &rpcdbpb.IteratorReleaseRequest{
+			Id: it.id,
+		})
+		if err != nil {
+			it.setError(err)
+		} else {
+			it.setError(errEnumToError[resp.Err])
+		}
+
+		close(it.fetchedData)
+		close(it.onClosed)
+	}()
+
+	for {
+		resp, err := it.db.client.IteratorNext(context.Background(), &rpcdbpb.IteratorNextRequest{
+			Id: it.id,
+		})
+		if err != nil {
+			it.setError(err)
+			return
+		}
+
+		if len(resp.Data) == 0 {
+			return
+		}
+
+		for {
+			select {
+			case it.fetchedData <- resp.Data:
+			case onUpdated := <-it.reqUpdateError:
+				it.updateError()
+				close(onUpdated)
+			case <-it.onClose:
+				return
+			}
+			break
+		}
+	}
 }
 
 // Next attempts to move the iterator to the next element and returns if this
 // succeeded
 func (it *iterator) Next() bool {
 	if it.db.closed.Get() {
```

Comment:
They iterators are not usable once db is closed, so this makes sense to me.

Reply:
So, the reason this isn't done is because this would require the `DatabaseClient` to actually track the outstanding `iterator`s... Because the iterator already needs a reference to the `DatabaseClient` this would introduce a kind of gross circular reference... I think keeping it as-is is probably the cleanest. We still require the iterator owner to call `Release` anyway according to the interface... So I think this is fine.